stingray/fourier.py

```diff
@@ -0,0 +1,1286 @@
+    if counts is None:
+        return generate_indices_of_segment_boundaries_unbinned
+
+    def fun(*args, **kwargs):
+        return generate_indices_of_segment_boundaries_binned(*args, dt=dt, **kwargs)
+    return fun
+
+
+def get_total_ctrate(times, gti, segment_size, counts=None):
+    """Calculate the average count rate during the observation.
+
+    This function finds the same segments that the PDS will use and
+    returns the mean count rate.
+    If ``counts`` is ``None``, the input times are interpreted as events.
+    Otherwise, the number of events is taken from ``counts``
+
+    Parameters
+    ----------
+    times : float `np.array`
+        Array of times
+    gti : [[gti00, gti01], [gti10, gti11], ...]
+        good time intervals
+    segment_size : float
+        length of segments
+
+    Other parameters
+    ----------------
+    counts : float `np.array`, default None
+        Array of counts per bin
+
+    Examples
+    --------
+    >>> times = np.sort(np.random.uniform(0, 1000, 1000))
+    >>> gti = np.asarray([[0, 1000]])
+    >>> counts, _ = np.histogram(times, bins=np.linspace(0, 1000, 11))
+    >>> bin_times = np.arange(50, 1000, 100)
+    >>> get_total_ctrate(bin_times, gti, 1000, counts=counts)
+    1.0
+    >>> get_total_ctrate(times, gti, 1000)
+    1.0
+    """
+    Nph = 0
+    Nintvs = 0
+    func = _which_segment_idx_fun(counts)
+
+    for _, _, idx0, idx1 in func(times, gti, segment_size):
+        if counts is None:
+            Nph += idx1 - idx0
+        else:
+            Nph += np.sum(counts[idx0:idx1])
+        Nintvs += 1
+
+    return Nph / (Nintvs * segment_size)
+
+
+def get_flux_iterable_from_segments(times, gti, segment_size, N=None, counts=None, errors=None):
+    """Get fluxes from different segments of the observation.
+
+    If ``counts`` is ``None``, the input times are interpreted as events.
+    At least one between ``N`` and ``counts`` needs to be specified.
+    Otherwise, they are assumed uniformly binned inside each GTI, and the number
+    of events per bin is taken from ``counts``
+
+    Parameters
+    ----------
+    times : float `np.array`
+        Array of times
+    gti : [[gti00, gti01], [gti10, gti11], ...]
+        good time intervals
+    segment_size : float
+        length of segments
+
+    Other parameters
+    ----------------
+    N : int, default None
+        Number of bins to divide the ``segment_size`` in
+    counts : float `np.array`, default None
+        Array of counts per bin
+
+    Returns
+    -------
+    cts : `np.array`
+        Array of counts
+    err : `np.array`
+        (optional) if ``errors`` is None, an array of errors in the segment
+
+    """
+    if counts is None and N is None:
+        raise ValueError(
+            "At least one between counts (if light curve) and N (if events) has to be set"
+        )
+
+    dt = None
+    if counts is not None:
+        dt = np.median(np.diff(times[:100]))
```

Comment:
Small potential breaking point: Fermi/GBM CTIME data changes time resolution during a trigger, which may or may not be within 100 bins of the start time, so this would fail for that type of data.

Reply:
But the assumption here is that the count array is uniformly sampled. If this is not true, there are _many_ other things that break here!

Reply:
Because the first dt can be, e.g. a single point before a GTI stops, and we do allow the sampling to be off outside GTI. It happens sometimes. Having 100 points, I'm reasonably sure that there will be enough points with the same sample time